grpc/src/main/java/io/stargate/grpc/service/interceptors/NewConnectionInterceptor.java

```diff
@@ -0,0 +1,128 @@
+import io.stargate.auth.UnauthorizedException;
+import io.stargate.db.AuthenticatedUser;
+import io.stargate.db.ClientInfo;
+import io.stargate.db.Persistence;
+import io.stargate.db.Persistence.Connection;
+import io.stargate.grpc.service.GrpcService;
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.Nullable;
+import org.immutables.value.Value;
+
+public class NewConnectionInterceptor implements ServerInterceptor {
+
+  public static final Metadata.Key<String> TOKEN_KEY =
+      Metadata.Key.of("X-Cassandra-Token", Metadata.ASCII_STRING_MARSHALLER);
+
+  private static final InetSocketAddress DUMMY_ADDRESS = new InetSocketAddress(9042);
+
+  private final Persistence persistence;
+  private final AuthenticationService authenticationService;
+  private final LoadingCache<RequestInfo, Connection> connectionCache =
+      Caffeine.newBuilder().expireAfterWrite(Duration.ofMinutes(1)).build(this::newConnection);
+
+  @Value.Immutable
+  interface RequestInfo {
+
+    String token();
+
+    Map<String, String> headers();
+
+    @Nullable
+    SocketAddress remoteAddress();
+  }
+
+  public NewConnectionInterceptor(
+      Persistence persistence, AuthenticationService authenticationService) {
+    this.persistence = persistence;
+    this.authenticationService = authenticationService;
+  }
+
+  @Override
+  public <ReqT, RespT> Listener<ReqT> interceptCall(
+      ServerCall<ReqT, RespT> call, Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+    try {
+      String token = headers.get(TOKEN_KEY);
+      if (token == null) {
+        call.close(Status.UNAUTHENTICATED.withDescription("No token provided"), new Metadata());
+        return new NopListener<>();
+      }
+
+      RequestInfo info =
+          ImmutableRequestInfo.builder()
+              .token(token)
+              .headers(convertAndFilterHeaders(headers))
+              .remoteAddress(call.getAttributes().get(Grpc.TRANSPORT_ATTR_REMOTE_ADDR))
+              .build();
+
+      Connection connection = connectionCache.get(info);
+
+      Context context = Context.current();
+      context = context.withValue(GrpcService.CONNECTION_KEY, connection);
+      return Contexts.interceptCall(context, call, headers, next);
+    } catch (Exception e) {
+      call.close(
+          Status.UNAUTHENTICATED.withDescription("Invalid token").withCause(e), new Metadata());
+    }
+    return new NopListener<>();
+  }
+
+  private Connection newConnection(RequestInfo info) throws UnauthorizedException {
+    AuthenticationSubject authenticationSubject = authenticationService.validateToken(info.token());
+
+    AuthenticatedUser user = authenticationSubject.asUser();
+    Connection connection;
+    if (!user.isFromExternalAuth()) {
+      SocketAddress remoteAddress = info.remoteAddress();
+      InetSocketAddress inetSocketAddress = DUMMY_ADDRESS;
```

Comment:
what's the deal with `DUMMY_ADDRESS` :smile: ?

Reply:
This is a best effort attempt to put the correct remote address in the `ClientState` object. If the remote address is not an `InetSocketAddress` I just put a dummy value in there instead of erroring out for something that's not critical. It will almost never not be a `InetSocketAddress` though.